Improved verification and checking to prevent possible 0x31 pod faults

itsmojo opened the pull request:
+ Add new non-persistent deliveryStatusVerified podState boolean
+ Validate podState insulin delivery state against each pod status return
+ Never schedule a basal if pod might possibly be not suspended
+ Never start a bolus if one might be possibly be running
+ Never start a temp basal if one might possibly be running
+ Add new PodCommsError for pod setup not complete condition
+ Add checks to never do a cancel command unless pod setup is complete
+ Use matching generic PumpManager error comment strings
+ Add unacknowledgedCommand checking to acknowledgeAlerts()
+ Add unacknowledgedCommand checking to configureAlerts()
+ Expanded and updated variable use comments

itsmojo committed:
Improved verification and checking to prevent possible 0x31 pod faults
+ Add new non-persistent deliveryStatusVerified podState boolean
+ Validate podState insulin delivery state against each pod status return
+ Never schedule a basal if pod might possibly be not suspended
+ Never start a bolus if one might be possibly be running
+ Never start a temp basal if one might possibly be running
+ Add new PodCommsError for pod setup not complete condition
+ Add checks to never do a cancel command unless pod setup is complete
+ Use matching generic PumpManager error comment strings
+ Add unacknowledgedCommand checking to acknowledgeAlerts()
+ Add unacknowledgedCommand checking to configureAlerts()

itsmojo committed:
Expanded and updated variable use comments

itsmojo committed:
Change deliveryStatusVerified state to deliveryStateInconsistencyDetected
+ Improved logic and comments for possible safety suspend before set basal
+ Updated comments for renamed non-persistent state with a reversed sense

marionbarker commented:
Sorry - pasted this into the wrong PR. Meant this for the OmniBLE PR 110.

~~Tested with both the previous version and the newly renamed variable with updated logic.~~
~~* iOS 16, using rPi DASH simulator~~
~~* In normal operation, the behavior is nominal~~

~~I'm not sure how to trigger a fault that would have caused an 0x31 to be sent in before making this modification.~~

marionbarker commented:
Tested this PR with both the previous version and the newly renamed variable with updated logic.
* iOS 17 attached to an Eros pod using a (Medronic antenna) rileylink
* In normal operation, the behavior is nominal

I'm not sure how to trigger a fault that would have caused an 0x31 to be sent in before making this modification.

itsmojo committed:
Defer configureAlerts & acknowledgeAlerts guards for some later PR

marionbarker commented:
Joe suggested a method for causing a failure of sending a command when the pod is not in the correct state to accept that command, which will cause an 0x31 fault.

I successfully used his idea to test this PR using an Eros pod with an SE phone, iOS 17.2.1. Code was at LoopWorkspace dev branch level with exceptions noted below.

## Summary:
* I tested first with code that has PR 24 applied and there were no failures (cancelTB or getStatus were sent before successive attempt to set TB or to issue a bolus)
* I then tested with the main branch (no changes from PR 24 applied) using this modification and got an immediate 0x31 fault
* I copied lines from the Xcode debug log file (filtered for the word "hex") into the reports posted below along with parsing for the send/recv hex patterns

## Method to cause a failure:

1. Comment out and add print statement preceding these lines: 
   * https://github.com/LoopKit/OmniKit/blob/df8d2fd/OmniKit/PumpManager/PodCommsSession.swift#L519
   * https://github.com/LoopKit/OmniKit/blob/df8d2fd/OmniKit/PumpManager/PodCommsSession.swift#L551

The print statements added before each commented out line are:
```
            print(" *hex* in bolus, podState.unfinalizedBolus not updated")
```
and
```
            print(" *hex* in setTempBasal, podState.unfinalizedTempBasal not updated")
```
respectively.

### Initial Test 

This was with PR 24 (before Joe's last commit to PR 24 - not a change that affects the results) applied to the same commit of OmniKit as is used by LoopWorkspace

[Xcode-Log-force-error-pr-24-only.txt](https://github.com/LoopKit/OmniKit/files/13944209/Xcode-Log-force-error-pr-24-only.txt)

### Final Test, 0x31 Fault

This was with OmniKit main (one commit ahead, but not a commit that affects the results)

[Xcode-Log-force-error.txt](https://github.com/LoopKit/OmniKit/files/13944212/Xcode-Log-force-error.txt)

itsmojo committed:
Use an optional lastDeliveryStatusReceived var for improved clarity
+ Updated commenting and variable testing to make things more obvious
+ Return noResponse for bolus check comms error instead of unfinalizedBolus
+ DeliveryStatus improvements to fix var name typo and add a suspended var
+ New optional initialDeliveryStatus parameter to PodState.init() for testing
+ Update OmniKit PodCommsSession tests to work with improved bolus checking

itsmojo committed:
Improved comment, add 2 missing protection guards

itsmojo committed:
Merge branch 'main' into insulin-delivery-checking

marionbarker commented:
Code review: 
* I compared OmniKit PR 24 with OmniBLE PR 110
* Both PR are consistent
* These PR add lastDeliveryStatusReceived as a double check to the internal state that Loop maintains for the pod
* It is a rare event for internal state to not match true pod state, e.g., [Loop Issue 2121](https://github.com/LoopKit/Loop/issues/2121)
* A few other cases where a command could fault the pod have double-checks added to prevent this unlikely mistake

Test with an Eros Pod running code that has OmniKit PR 24 applied to the code.
* The test phone was fed with a glucose pattern that stimulated Loop to modify dosing
* Both automatic bolus and temp basal only dosing methods were applied
* Loop behaved nominally over many hours
* I also confirmed that a new Eros pod goes through the pairing process as expected

I did a few test cases:

Standard code (with PR)
Case 1: Automatic temp basal is issued when the internal state indicates that scheduled basal is running
* Expect Loop to send the setTB command only
* I confirmed this to be true

Standard code (with PR)
Case 2: Automatic temp basal is issued when the internal state indicates that temporary basal is running
* Expect Loop to send the cancelTB command, get a response, and then send the setTB command
* I confirmed this to be true

I then applied a test patch that overwrites the lastDeliveryStatusReceived with a flag indicating a temporary basal is running:

Standard code (with PR and test patch)
Case 3: Automatic temp basal is issued when the internal state indicates that scheduled basal is running
* Loop uses the extra logic check with lastDeliveryStatusReceived
   * Because this check fails (with this patch), an extra cancelTB is sent
* Expect Loop to send the cancelTB command, get a response, send the setTB command
* I confirmed this to be true

Standard code (with PR and test patch)
Case 4: Update scheduled basal program
* This always starts with a cancelAll to suspend pod
* New check adds lastDeliveryStatusReceived being in suspended state in addition to internal state
   * Because this check fails (with this patch), an extra cancelAll is sent
   * I confirmed this is true

```
Loop Sent: CancelAll
Pod  Sent:  Basal suspended
Loop Sent: CancelAll
Pod  Sent:  Basal suspended
Loop Sent: Modify Basal Schedule
```

Restored to Standard code (with PR):
Case 5: Update scheduled basal program
* Observed nominal sequence

```
Loop Sent: CancelAll
Pod  Sent:  Basal suspended
Loop Sent: Modify Basal Schedule
```
